tsdb/head_append.go

```diff
@@ -1012,411 +1028,472 @@ func (a *headAppender) Commit() (err error) {
 	for i, m := range a.metadata {
-		series = a.metadataSeries[i]
-		series.Lock()
-		series.meta = &metadata.Metadata{Type: record.ToMetricType(m.Type), Unit: m.Unit, Help: m.Help}
-		series.Unlock()
+		acc.series = a.metadataSeries[i]
+		acc.series.Lock()
+		acc.series.meta = &metadata.Metadata{Type: record.ToMetricType(m.Type), Unit: m.Unit, Help: m.Help}
+		acc.series.Unlock()
+	}
+}
+
+// Commit writes to the WAL and adds the data to the Head.
+// TODO(codesome): Refactor this method to reduce indentation and make it more readable.
+func (a *headAppender) Commit() (err error) {
+	if a.closed {
+		return ErrAppenderClosed
+	}
+	defer func() { a.closed = true }()
+
+	if err := a.log(); err != nil {
+		_ = a.Rollback() // Most likely the same error will happen again.
+		return fmt.Errorf("write to WAL: %w", err)
+	}
+
+	if a.head.writeNotified != nil {
+		a.head.writeNotified.Notify()
 	}
 
-	a.head.metrics.outOfOrderSamples.WithLabelValues(sampleMetricTypeFloat).Add(float64(floatOOORejected))
-	a.head.metrics.outOfOrderSamples.WithLabelValues(sampleMetricTypeHistogram).Add(float64(histoOOORejected))
-	a.head.metrics.outOfBoundSamples.WithLabelValues(sampleMetricTypeFloat).Add(float64(floatOOBRejected))
-	a.head.metrics.tooOldSamples.WithLabelValues(sampleMetricTypeFloat).Add(float64(floatTooOldRejected))
-	a.head.metrics.samplesAppended.WithLabelValues(sampleMetricTypeFloat).Add(float64(floatsAppended))
-	a.head.metrics.samplesAppended.WithLabelValues(sampleMetricTypeHistogram).Add(float64(histogramsAppended))
-	a.head.metrics.outOfOrderSamplesAppended.WithLabelValues(sampleMetricTypeFloat).Add(float64(oooFloatsAccepted))
-	a.head.metrics.outOfOrderSamplesAppended.WithLabelValues(sampleMetricTypeHistogram).Add(float64(oooHistogramAccepted))
-	a.head.updateMinMaxTime(inOrderMint, inOrderMaxt)
-	a.head.updateMinOOOMaxOOOTime(oooMinT, oooMaxT)
+	a.commitExemplars()
+
+	defer a.head.metrics.activeAppenders.Dec()
+	defer a.head.putAppendBuffer(a.samples)
+	defer a.head.putSeriesBuffer(a.sampleSeries)
+	defer a.head.putExemplarBuffer(a.exemplars)
+	defer a.head.putHistogramBuffer(a.histograms)
+	defer a.head.putFloatHistogramBuffer(a.floatHistograms)
+	defer a.head.putMetadataBuffer(a.metadata)
+	defer a.head.iso.closeAppend(a.appendID)
+
+	acc := &appenderCommitContext{
+		floatsAppended:     len(a.samples),
+		histogramsAppended: len(a.histograms) + len(a.floatHistograms),
+		inOrderMint:        math.MaxInt64,
+		inOrderMaxt:        math.MinInt64,
+		oooMinT:            math.MaxInt64,
+		oooMaxT:            math.MinInt64,
+		oooCapMax:          a.head.opts.OutOfOrderCapMax.Load(),
+		appendChunkOpts: chunkOpts{
+			chunkDiskMapper: a.head.chunkDiskMapper,
+			chunkRange:      a.head.chunkRange.Load(),
+			samplesPerChunk: a.head.opts.SamplesPerChunk,
+		},
+	}
+
+	defer func() {
+		for i := range acc.oooRecords {
+			a.head.putBytesBuffer(acc.oooRecords[i][:0])
+		}
+	}()
 
-	collectOOORecords()
+	a.commitSamples(acc)
```

Comment:
I wasn't sure if we gain any benefit on run these commit functions in parallel, I tested dispatching one go routine for each one, tests were working as expected, but not sure.

I'm open for feedback.

Reply:
We can play with this. idea in a separate PR and with a benchmark.

In theory these should not block on each other so I can see the argument for running them in parallel.

Reply:
I think a problem with running these in parallel is that you are not deterministic anymore with the behavior when a series receives a float and histogram sample as well. This is rare but can happen.

In fact I'd rather merge the 3 loops into one and actually write the samples (floats and histograms) in the order they were appended and not always in the floats then histograms order. See #15142